### PR TITLE
Add ProtocolBehaviour

### PR DESCRIPTION
**Pull request by tomaka**
The next step in the refactoring is to implement the libp2p `NetworkBehaviour` trait on `Protocol`. This was done by adding a `ProtocolBehaviour` struct.

In other words:
Before: `CustomProto <-> Service <-> Protocol`.
After: `Service <-> (CustomProto <-> Protocol)`, and the thing in parentheses is `ProtocolBehaviour`.

This made it possible to move `start_service` in the constructor of `NetworkWorker`, which I did as well.

This unlocks the next step, which is to split `Protocol` in multiple `NetworkBehaviour`s.

Because #2500 was "rejected", I've had to use some hacks for granting access to the gossiping and specialization. It is unclear how I will clean that up, but I'll try to find a solution.

### CODE REVIEW

File: core/network/src/protocol_behaviour.rs [hunk 1]
```diff
@@ -0,0 +1,463 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
// Copyright 2019 Parity Technologies (UK) Ltd.
```

File: core/network/src/protocol_behaviour.rs [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+		self.protocol.gossip_consensus_message(
+			&mut LocalNetworkOut { inner: &mut self.behaviour, peerset_handle: &self.peerset_handle },
+			topic,
+			engine_id,
+			message,
+			recipient
+		);
+	}
+
+	// TODO: needed?
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]: Every `TODO` should be an issue in github.

File: core/network/src/protocol_behaviour.rs [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+		self.protocol.clear_justification_requests()
+	}
+
+	/// A batch of blocks have been processed, with or without errors.
+	/// Call this when a batch of blocks have been processed by the import queue, with or without
+	/// errors.
+	pub fn blocks_processed(
+		&mut self,
+		processed_blocks: Vec<B::Hash>,
+		has_error: bool
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
		has_error: bool,
```

File: core/network/src/protocol_behaviour.rs [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+	) {
+		self.protocol.blocks_processed(
+			&mut LocalNetworkOut { inner: &mut self.behaviour, peerset_handle: &self.peerset_handle },
+			processed_blocks,
+			has_error
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
			has_error,
```

File: core/network/src/protocol_behaviour.rs [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+		self.protocol.justification_import_result(hash, number, success)
+	}
+
+	/// Request a finality proof for the given block.
+	///
+	/// Queues a new finality proof request and tries to dispatch all pending requests.
+	pub fn request_finality_proof(
+		&mut self,
+		hash: &B::Hash,
+		number: NumberFor<B>
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
		number: NumberFor<B>,
```

File: core/network/src/protocol_behaviour.rs [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
+	) {
+		self.protocol.request_finality_proof(
+			&mut LocalNetworkOut { inner: &mut self.behaviour, peerset_handle: &self.peerset_handle },
+			&hash,
+			number
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
			number,
```

File: core/network/src/protocol_behaviour.rs [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+	}
+
+	fn inject_disconnected(&mut self, peer_id: &PeerId, endpoint: ConnectedPoint) {
+		self.behaviour.inject_disconnected(peer_id, endpoint)
+	}
+
+	fn inject_node_event(
+		&mut self,
+		peer_id: PeerId,
+		event: <<Self::ProtocolsHandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::OutEvent
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
		event: <<Self::ProtocolsHandler as IntoProtocolsHandler>::Handler as ProtocolsHandler>::OutEvent,
```

File: core/network/src/protocol_behaviour.rs [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
+	) {
+		self.behaviour.inject_node_event(peer_id, event)
+	}
+
+	fn poll(
+		&mut self,
+		params: &mut PollParameters
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
		params: &mut PollParameters,
```

File: core/network/src/protocol_behaviour.rs [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+				return Async::Ready(NetworkBehaviourAction::DialPeer { peer_id }),
+			Async::Ready(NetworkBehaviourAction::SendEvent { peer_id, event }) =>
+				return Async::Ready(NetworkBehaviourAction::SendEvent { peer_id, event }),
+			Async::Ready(NetworkBehaviourAction::ReportObservedAddr { address }) =>
+				return Async::Ready(NetworkBehaviourAction::ReportObservedAddr { address }),
+		};
+
+		let mut network_out = LocalNetworkOut {
+			inner: &mut self.behaviour,
+			peerset_handle: &self.peerset_handle
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
			peerset_handle: &self.peerset_handle,
```

File: core/network/src/protocol_behaviour.rs [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+	fn inject_expired_listen_addr(&mut self, addr: &Multiaddr) {
+		self.behaviour.inject_expired_listen_addr(addr)
+	}
+
+	fn inject_new_external_addr(&mut self, addr: &Multiaddr) {
+		self.behaviour.inject_new_external_addr(addr)
+	}
+}
+
+impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> DiscoveryNetBehaviour for
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> DiscoveryNetBehaviour
```

File: core/network/src/protocol_behaviour.rs [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+ProtocolBehaviour<B, S, H> {
```
Comment (bkchr) [inline core/network/src/protocol_behaviour.rs]:
```suggestion
	for ProtocolBehaviour<B, S, H> {
```

**Comment by tomaka** [comment]
Fixed the nitpicks!

**Pull request merged**